Panic in TestDeltaProxyCacheRequestPartialHitChunks

We're seeing this Panic occasionally when running tests. I found that setting my computer clock to 4:15 EDT Mar 30 2025 will cause this test (and _only_ this test) to fail. Poking around shows that in the earlier goroutine, if `outIdx` is 0, `msgp` will return an error trying to decode the cache entry when the computer clock is around this time. Currently the error is not logged, which is why this was missed. Because the Panic is time-of-day-dependent, it suggests a possible defect affecting the caching and retrieval of the first chunk for the aggregated response based on the time of day.

To reproduce, rather than changing your local clock, you can temporarily modify this test at `pkg/proxy/engines/delta_proxy_cache_chunk_test.go:338` as follows:

```
	// now := time.Now()
	now := time.Unix(1743365100, 0) // TODO: REMOVE
```


Failure Condition:
```
--- FAIL: TestDeltaProxyCacheRequestPartialHitChunks (0.03s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x40 pc=0x10353937c]

goroutine 286 [running]:
testing.tRunner.func1.2({0x1039a3020, 0x1042ccfa0})
	/go/1.24.1/libexec/src/testing/testing.go:1734 +0x1ac
testing.tRunner.func1()
	/go/1.24.1/libexec/src/testing/testing.go:1737 +0x334
panic({0x1039a3020?, 0x1042ccfa0?})
	/go/1.24.1/libexec/src/runtime/panic.go:792 +0x124
github.com/trickstercache/trickster/v2/pkg/proxy/engines.QueryCache({0x103b161b8, 0x140007620c0}, {0x103b1ec48, 0x140000ee9a0}, {0x14000036280, 0x34}, {0x0, 0x0, 0x0}, 0x103b03428)
	/trickster/pkg/proxy/engines/cache.go:189 +0xe1c
github.com/trickstercache/trickster/v2/pkg/proxy/engines.DeltaProxyCacheRequest({0x103b141e0, 0x140006a0400}, 0x140006a63c0, 0x14000762090)
	/trickster/pkg/proxy/engines/deltaproxycache.go:145 +0xb08
github.com/trickstercache/trickster/v2/pkg/proxy/engines.(*TestClient).QueryRangeHandler(0x989680?, {0x103b141e0, 0x140006a0400}, 0x140006a63c0)
	/trickster/pkg/proxy/engines/client_test.go:378 +0xbc
github.com/trickstercache/trickster/v2/pkg/proxy/engines.TestDeltaProxyCacheRequestPartialHitChunks(0x140005a68c0)
	/trickster/pkg/proxy/engines/deltaproxycache_chunk_test.go:432 +0x1198
testing.tRunner(0x140005a68c0, 0x103b02ef0)
	/go/1.24.1/libexec/src/testing/testing.go:1792 +0xe4
created by testing.(*T).Run in goroutine 1
	/go/1.24.1/libexec/src/testing/testing.go:1851 +0x374
exit status 2
FAIL	github.com/trickstercache/trickster/v2/pkg/proxy/engines	0.544s
```

--- a/go.mod
+++ b/go.mod
@@ -2,7 +2,7 @@ module github.com/trickstercache/trickster/v2
 
 go 1.24
 
-toolchain go1.24.0
+toolchain go1.24.1
 
 require (
 	github.com/alicebob/miniredis v2.5.0+incompatible

--- a/pkg/observability/logging/logging.go
+++ b/pkg/observability/logging/logging.go
@@ -336,8 +336,12 @@ func (l *logger) log(logLevel level.Level, event string, detail Pairs) {
 	if ld > 0 {
 		sortedKeys := make([]string, ld)
 		for k, v := range detail {
-			if s, ok := v.(string); ok && strings.Contains(s, " ") {
-				v = `"` + s + `"`
+			if s, ok := v.(string); ok {
+				v = quoteAsNeeded(s)
+			} else if stringer, ok := v.(fmt.Stringer); ok {
+				v = quoteAsNeeded(stringer.String())
+			} else if err, ok := v.(error); ok {
+				v = quoteAsNeeded(err.Error())
 			}
 			sortedKeys[i] = fmt.Sprintf("%s=%v", k, v)
 			i++

--- a/pkg/proxy/engines/cache.go
+++ b/pkg/proxy/engines/cache.go
@@ -162,7 +162,7 @@ func QueryCache(ctx context.Context, c cache.Cache, key string,
 			// Prepare buffered results and waitgroup
 			wg := &sync.WaitGroup{}
 			// Result slice of timeseries
-			ress := make([]timeseries.Timeseries, cct)
+			ress := make(timeseries.TimeseriesList, cct)
 			resi := 0
 			for chunkStart := cext.Start; chunkStart.Before(cext.End); chunkStart = chunkStart.Add(csize) {
 				// Chunk range (inclusive, on-step)
@@ -204,27 +204,9 @@ func QueryCache(ctx context.Context, c cache.Cache, key string,
 			}
 			// Wait on queries
 			wg.Wait()
-			// this removes any nil entries (cache misses) from ress
-			cts := make([]timeseries.Timeseries, len(ress))
-			var j int
-			for _, ts := range ress {
-				if ts == nil {
-					continue
-				}
-				cts[j] = ts
-				j++
-			}
-			ress = cts[:j]
-			// this merges all of the timeseries from the different cache subkeys
-			// into a unified single timeseries
-			if j > 0 {
-				d.timeseries = ress[0]
-				if j > 1 {
-					d.timeseries.Merge(true, ress[1:]...)
-				}
-				if d.timeseries != nil {
-					d.timeseries.SetExtents(d.timeseries.Extents().Compress(trq.Step))
-				}
+			d.timeseries = ress.Merge()
+			if d.timeseries != nil {
+				d.timeseries.SetExtents(d.timeseries.Extents().Compress(trq.Step))
 			}
 		} else {
 			// Do byterange chunking

--- a/pkg/proxy/engines/cache.go
+++ b/pkg/proxy/engines/cache.go
@@ -181,15 +181,23 @@ func QueryCache(ctx context.Context, c cache.Cache, key string,
 						(qr.err == nil || qr.err == cache.ErrKNF) {
 						return
 					}
-					if c.Configuration().Provider != "memory" {
-						qr.d.timeseries, qr.err = unmarshal(qr.d.Body, nil)
-					}
-					if qr.err == nil && qr.d != nil && qr.d.timeseries != nil {
-						ress[outIdx] = qr.d.timeseries
-					} else {
+					if qr.err != nil {
 						logger.Error("dpc query cache chunk failed",
 							logging.Pairs{"error": qr.err, "chunkIdx": outIdx,
 								"key": subkey, "cacheQueryStatus": qr.lookupStatus})
+						return
+					}
+					if c.Configuration().Provider != "memory" {
+						qr.d.timeseries, qr.err = unmarshal(qr.d.Body, nil)
+						if qr.err != nil {
+							logger.Error("dpc query cache chunk failed",
+								logging.Pairs{"error": qr.err, "chunkIdx": outIdx,
+									"key": subkey, "cacheQueryStatus": qr.lookupStatus})
+							return
+						}
+					}
+					if qr.d.timeseries != nil {
+						ress[outIdx] = qr.d.timeseries
 					}
 				}(resi)
 				resi++

--- a/pkg/proxy/engines/cache.go
+++ b/pkg/proxy/engines/cache.go
@@ -20,6 +20,7 @@ import (
 	"bytes"
 	"context"
 	"errors"
+	"fmt"
 	"io"
 	"mime"
 	"net/http"
@@ -170,30 +171,52 @@ func QueryCache(ctx context.Context, c cache.Cache, key string,
 					End:   chunkStart.Add(csize - trq.Step),
 				}
 				// Derive subkey
-				subkey := key + chunkExtent.String()
+				subkey := getSubKey(key, chunkExtent)
 				// Query
 				wg.Add(1)
 				go func(outIdx int) {
 					defer wg.Done()
 					qr := queryConcurrent(ctx, c, subkey, nil, nil)
+					if qr.lookupStatus != status.LookupStatusHit &&
+						(qr.err == nil || qr.err == cache.ErrKNF) {
+						return
+					}
 					if c.Configuration().Provider != "memory" {
 						qr.d.timeseries, qr.err = unmarshal(qr.d.Body, nil)
 					}
-					if qr.err == nil {
+					if qr.err == nil && qr.d != nil && qr.d.timeseries != nil {
 						ress[outIdx] = qr.d.timeseries
 					} else {
 						logger.Error("dpc query cache chunk failed",
-							logging.Pairs{"detail": qr.err, "chunkIdx": outIdx})
+							logging.Pairs{"error": qr.err, "chunkIdx": outIdx,
+								"key": subkey, "cacheQueryStatus": qr.lookupStatus})
 					}
 				}(resi)
 				resi++
 			}
 			// Wait on queries
 			wg.Wait()
-			d.timeseries = ress[0]
-			d.timeseries.Merge(true, ress[1:]...)
-			if d.timeseries != nil {
-				d.timeseries.SetExtents(d.timeseries.Extents().Compress(trq.Step))
+			// this removes any nil entries (cache misses) from ress
+			cts := make([]timeseries.Timeseries, len(ress))
+			var j int
+			for _, ts := range ress {
+				if ts == nil {
+					continue
+				}
+				cts[j] = ts
+				j++
+			}
+			ress = cts[:j]
+			// this merges all of the timeseries from the different cache subkeys
+			// into a unified single timeseries
+			if j > 0 {
+				d.timeseries = ress[0]
+				if j > 1 {
+					d.timeseries.Merge(true, ress[1:]...)
+				}
+				if d.timeseries != nil {
+					d.timeseries.SetExtents(d.timeseries.Extents().Compress(trq.Step))
+				}
 			}
 		} else {
 			// Do byterange chunking
@@ -431,7 +454,7 @@ func WriteCache(ctx context.Context, c cache.Cache, key string, d *HTTPDocument,
 					End:   chunkStart.Add(csize - trq.Step),
 				}
 				// Derive subkey
-				subkey := key + chunkExtent.String()
+				subkey := getSubKey(key, chunkExtent)
 				// Query
 				wg.Add(1)
 				go func() {
@@ -564,4 +587,8 @@ func DocumentFromHTTPResponse(resp *http.Response, body []byte,
 	}
 
 	return d
+}
+
+func getSubKey(key string, chunkExtent timeseries.Extent) string {
+	return fmt.Sprintf("%s.%s", key, chunkExtent)
 }

--- a/pkg/proxy/engines/cache.go
+++ b/pkg/proxy/engines/cache.go
@@ -29,6 +29,8 @@ import (
 
 	"github.com/trickstercache/trickster/v2/pkg/cache"
 	"github.com/trickstercache/trickster/v2/pkg/cache/status"
+	"github.com/trickstercache/trickster/v2/pkg/observability/logging"
+	"github.com/trickstercache/trickster/v2/pkg/observability/logging/logger"
 	tspan "github.com/trickstercache/trickster/v2/pkg/observability/tracing/span"
 	tc "github.com/trickstercache/trickster/v2/pkg/proxy/context"
 	"github.com/trickstercache/trickster/v2/pkg/proxy/headers"
@@ -179,6 +181,9 @@ func QueryCache(ctx context.Context, c cache.Cache, key string,
 					}
 					if qr.err == nil {
 						ress[outIdx] = qr.d.timeseries
+					} else {
+						logger.Error("dpc query cache chunk failed",
+							logging.Pairs{"detail": qr.err, "chunkIdx": outIdx})
 					}
 				}(resi)
 				resi++

--- a/pkg/timeseries/dataset/dataset_test.go
+++ b/pkg/timeseries/dataset/dataset_test.go
@@ -22,6 +22,7 @@ import (
 	"testing"
 	"time"
 
+	"github.com/trickstercache/trickster/v2/pkg/observability/logging/logger"
 	"github.com/trickstercache/trickster/v2/pkg/timeseries"
 	"github.com/trickstercache/trickster/v2/pkg/timeseries/epoch"
 )
@@ -42,7 +43,7 @@ func testDataSet() *DataSet {
 
 func genTestDataSet(seriesCount int, resultsCount int) *DataSet {
 	if resultsCount <= 0 {
-		fmt.Println("resultsCount must be greater than 0")
+		logger.Error("resultsCount must be greater than 0", nil)
 		resultsCount = 1
 	}
 	results := make([]*Result, 0, resultsCount)

--- a/pkg/timeseries/timeseries.go
+++ b/pkg/timeseries/timeseries.go
@@ -80,3 +80,36 @@ type Timeseries interface {
 	// SetTimeRangeQuery sets the TimeRangeQuery associated with the Timeseries
 	SetTimeRangeQuery(*TimeRangeQuery)
 }
+
+type TimeseriesList []Timeseries
+
+// Compress removes nil entries from t
+func (t TimeseriesList) Compress() TimeseriesList {
+	if len(t) == 0 {
+		return TimeseriesList{}
+	}
+	out := make(TimeseriesList, len(t))
+	var j int
+	for _, ts := range t {
+		if ts == nil {
+			continue
+		}
+		out[j] = ts
+		j++
+	}
+	return out[:j]
+}
+
+// Merge combines all Timesseries in t into a single Timeseries
+func (t TimeseriesList) Merge() Timeseries {
+	cts := t.Compress()
+	if len(cts) == 0 {
+		return nil
+	}
+	if len(cts) == 1 {
+		return cts[0]
+	}
+	out := cts[0].Clone()
+	out.Merge(true, cts[1:]...)
+	return out
+}

